src/encoder.rs

```diff
@@ -79,27 +79,69 @@ impl Frame {
         }
     }
 
-    /// Returns a `Vec` containing the data of this frame's planes in YUV444 format.
-    /// Each point in the `Vec` is a triple consisting of a Y, U, and V component.
-    /// The `Vec` is laid out as contiguous rows, e.g. to get a given 0-indexed row
-    /// you could use `data[(width * row_idx)..(width * (row_idx + 1))]`.
+    /// Returns a `PixelIter` containing the data of this frame's planes in YUV format.
+    /// Each point in the `PixelIter` is a triple consisting of a Y, U, and V component.
+    /// The `PixelIter` is laid out as contiguous rows, e.g. to get a given 0-indexed row
+    /// you could use `data.skip(width * row_idx).take(width)`.
     ///
     /// This data retains any padding, e.g. it uses the width and height specifed in
     /// the Y-plane's `cfg` struct, and not the display width and height specied in
     /// `FrameInvariants`.
-    pub fn as_yuv444(&self) -> Vec<(u16, u16, u16)> {
-      let (width, height) = (self.planes[0].cfg.width, self.planes[0].cfg.height);
-      let mut result = Vec::with_capacity(width * height);
-      for y in 0..height {
-        let y_slice = self.planes[0].slice(&PlaneOffset { x: 0, y: y as isize });
-        let u_slice = self.planes[1].slice(&PlaneOffset { x: 0, y: y as isize / 2 });
-        let v_slice = self.planes[2].slice(&PlaneOffset { x: 0, y: y as isize / 2 });
-        for x in 0..width {
-          result.push((y_slice.p(x, 0), u_slice.p(x / 2, 0), v_slice.p(x / 2, 0)));
-        }
-      }
-      result
+    pub fn iter(&self) -> PixelIter {
+      PixelIter::new(&self.planes)
+    }
+}
+
+#[derive(Debug)]
+pub struct PixelIter<'a> {
+  planes: &'a [Plane; 3],
+  y: usize,
+  x: usize,
+}
+
+impl<'a> PixelIter<'a> {
+  pub fn new(planes: &'a [Plane; 3]) -> Self {
+    PixelIter {
+      planes,
+      y: 0,
+      x: 0,
     }
+  }
+
+  fn width(&self) -> usize {
+    self.planes[0].cfg.width
+  }
+
+  fn height(&self) -> usize {
+    self.planes[0].cfg.height
+  }
+}
+
+impl<'a> Iterator for PixelIter<'a> {
+  type Item = (u16, u16, u16);
+
+  fn next(&mut self) -> Option<<Self as Iterator>::Item> {
+    if self.y == self.height() - 1 && self.x == self.width() - 1 {
+      return None;
+    }
+    let pixel = (
+      self.planes[0].p(self.x, self.y),
+      self.planes[0].p(self.x / 2, self.y / 2),
```

Comment:
Looks strange, the previous code was using all the planes.

Reply:
Whoops, copy and paste error.